cpp/src/umap/simpl_set_embed/algo.h

```diff
@@ -96,70 +95,70 @@ __device__ __host__ T clip(T val, T lb, T ub) {
 }
 
 /**
-	         * Calculate the repulsive gradient
-	         */
-template <typename T>
-__device__ __host__ T repulsive_grad(T dist_squared, float gamma,
-                                     UMAPParams params) {
-  T grad_coeff = 2.0 * gamma * params.b;
+ * Calculate the repulsive gradient
+ */
+__device__ __host__ double repulsive_grad(double dist_squared, double gamma,
+                                          UMAPParams params) {
+  double grad_coeff = 2.0 * gamma * params.b;
   grad_coeff /=
     (0.001 + dist_squared) * (params.a * pow(dist_squared, params.b) + 1);
   return grad_coeff;
 }
 
 /**
-	         * Calculate the attractive gradient
-	         */
-template <typename T>
-__device__ __host__ T attractive_grad(T dist_squared, UMAPParams params) {
-  T grad_coeff = -2.0 * params.a * params.b * pow(dist_squared, params.b - 1.0);
+  * Calculate the attractive gradient
+  */
+__device__ __host__ double attractive_grad(double dist_squared,
+                                           UMAPParams params) {
+  double grad_coeff =
+    -2.0 * params.a * params.b * pow(dist_squared, params.b - 1.0);
   grad_coeff /= params.a * pow(dist_squared, params.b) + 1.0;
   return grad_coeff;
 }
 
 /**
-	         * Kernel for performing 1 epoch of stochastic gradient descent
-	         * on each call. Vectors are sampled in proportion to their
-	         * weights in the 1-skeleton. Negative samples are drawn
-	         * randomly.
-	         */
+ * Kernel for performing 1 epoch of stochastic gradient descent
+ * on each call. Vectors are sampled in proportion to their
+ * weights in the 1-skeleton. Negative samples are drawn
+ * randomly.
+ */
 template <typename T, int TPB_X>
 __global__ void optimize_batch_kernel(
   T *head_embedding, int head_n, T *tail_embedding, int tail_n, const int *head,
```

Comment:
Another custom kernel.

Reply:
I don’t believe using custom kernels in algorithms is bad if they aren’t reusable enough to use the prims or to become their own prim. In many of these cases with umap, I’ve used prims as much as possible but also want to keep the custom code like this matched with the umap reference implementation so that it is easy to follow.

Reply:
I am not saying make this function ml-prims. What I am saying that you may use the existing prims mostly rather than writing custom kernels.

Reply:
I am not saying make this function ml-prims. What I am saying that you may use the existing prims mostly rather than writing custom kernels.